### PR TITLE
Order Creation: Adds OrderSynchronizer Protocol

### LINKED ISSUE #6127: Order Creation  - Product Sync: Create Skeleton + protocol contract
We need to keep the local product in sync with the backend in order to properly calculate taxes. 

As the amount of work is considerable we should begin by creating a skeleton with a proper API contract so the consumers can continue with development.

Features to support that should be included in the contract:

    - publisher for loading
    - publisher for errors
    - publisher for values/order
    - add product
    - update product quantity
    - remove product
    - change status
    - add customer details
    - add shipping method
    - remove shipping method
    - update shipping method
    - add fee
    - remove fee
    - update fee 
    - create order

### PR DESCRIPTION
**Pull request by Ecarrion**
Closes: #6127 

# Why

This PR adds an `OrderSynchronizer` protocol. 

The idea behind this protocol is for us to quickly provide a simple `LocalOrderSynchronizer` so the `NewOrderViewModel` can continue its development with a real interface while the real order synchronizer is built.

# Testing Steps

- Make sure the interface looks correct and we are not missing any inputs or outputs.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by Ecarrion** [comment]
@rachelmcr seems like `AddProductToOrderViewModelTests.test_searching_products_returns_expected_results()` is failing on CI, do you know what is happening?

**Comment by rachelmcr** [comment]
@Ecarrion Ugh, thanks for the ping; that test is flaky. I thought I'd fixed it, but I got it to fail again locally after quite a few tries. I think I understand now why it was failing and have a fix for it in https://github.com/woocommerce/woocommerce-ios/pull/6153.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/OrderSynchronizer.swift [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    let product: Product
+    let quantity: Int
+}
+
+/// Addresses input for an `OrderSynchronizer` type.
+///
+struct OrderSyncAddressesInput {
+    let billing: Address
+    let shipping: Address
+}
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/OrderSynchronizer.swift:31]: Can we replace it with the existing `NewOrderAddressData`? There are optional properties, but it's easy to change.

https://github.com/woocommerce/woocommerce-ios/blob/820ca1f6d408702155aa44ecc8ed96e75df5079a/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/CustomerSection/CreateOrderAddressFormViewModel.swift#L7-L10

**Review by ealeksandrov — APPROVED** [review]
LGTM :shipit:
As we discussed, feel free to change some properties types now or in later PRs.

**Pull request merged**